dev-refactor merge

felsweg-iota opened the pull request:
# Description of change

This is the branch for upcoming features. No explicit change, but to keep `dev-refactor` mergable with `dev`. Merge this PR into `dev`, when all other changes are completed. 

## Type of change

- [x] Enhancement (a non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Fix

## How the change has been tested

All dependent branches and features have thoroughly been tested by unit and integration tests. 

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

felsweg-iota committed:
CI: add 'dev-refactor' branch for upcoming changes; add branch for CI to run; fix clippy errors

felsweg-iota committed:
CI: add 'dev-refactor' branch for upcoming changes; add branch for CI to run; fix clippy errors

felsweg-iota committed:
CI: add 'dev-refactor' branch for upcoming changes; add branch for CI to run; fix clippy errors

felsweg-iota committed:
CI: add 'dev-refactor' branch for upcoming changes; add branch for CI to run; fix clippy errors

felsweg-iota committed:
CI: add 'dev-refactor' branch for upcoming changes; add branch for CI to run; fix clippy errors

Commit:
Set architecture for new runtime, add ArrayVec crate

Commit:
Remove depedency to ArrayVec

Commit:
Implementation of Buffer type

Commit:
Add a README.md

Commit:
Restructure project

Commit:
Implementation of EncryptedRam

Commit:
Tests for EncryptedRam

Commit:
Basic implementation of FileMemory

Commit:
Implement Zeroize

Commit:
Format the code with rustfmt

Commit:
Test for FileMemory

Commit:
Add Encryption to FileMemory

Commit:
Zeroize on Drop

Commit:
Develop first iteration of non-contiguous memory

Commit:
Add refresh in non contiguous memory

vuongDang committed:
Correct typos

Co-authored-by: Matthias Kandora <matthias.kandora@iota.org>
Co-authored-by: Elena Frank <elena.frank@protonmail.com>

Commit:
Replace 'cypher' with 'cipher'

elenaf9 commented on .github/workflows/covector-version-or-publish.yml:
```diff
@@ -10,6 +10,7 @@ on:
   push:
     branches:
       - dev
+      - dev-refactor
```
Why do we need the covector workflow for dev-refactor?

felsweg-iota replied:
Agreed. That shouldn't be part of `dev-refactor`.

felsweg-iota committed:
fix: remove dev-refactor from covector version or publish

felsweg-iota committed:
fix: remove dev-refactor from covector version or publish

felsweg-iota committed:
fix: remove dev-refactor from covector version or publish

felsweg-iota committed:
fix: remove dev-refactor from covector version or publish

Commit:
Chage allocation and configuration of LockedMemory types

Commit:
Simplify the API

Commit:
Finalize zeroizing memories

Commit:
Format code with clippy and prehook

Commit:
Add cloning to the memories

Commit:
Finalize API

Commit:
Add intergration tests for LockedMemory

Commit:
Test cloning of locked memories

Commit:
Add access control for FileMemory

Commit:
Update the README.md

Commit:
Secure file data with noise

Commit:
Store files in local data directory

Commit:
Add zeroize test for Boxed type

Commit:
Add common implementation for Debug and tests

Commit:
Add forgotten staged files in previous commit

vuongDang committed:
Replace old runtime GuardedVec with new_runtime Buffer

vuongDang committed:
Remove crypto from the runtime

vuongDang committed:
Add non contiguous key type NCKey in crypto box

vuongDang committed:
Merge with dev-refactor

vuongDang committed:
Remove bytes() function from crypto_box as opens a security flaw

vuongDang committed:
Add serialization for new runtime types

vuongDang committed:
Add Encrypted Keystore in the engine crate

vuongDang committed:
Merge feat/locked-memory into dev-refactor

vuongDang committed:
Refactor with clippy and fmt

vuongDang committed:
Make file memory compatible to non-unix

vuongDang committed:
Format with clippy and fmt

elenaf9 commented on engine/new_runtime/src/memories/mod.rs:
```diff
@@ -0,0 +1,7 @@
+// Copyright 2020-2021 IOTA Stiftung
```
Nit: Move the mod file from `src/memories/mod.rs` -> `src/memories.rs` to be consistent in the module file naming throughout the crate.

elenaf9 commented on engine/new_runtime/src/memories/buffer.rs:
```diff
@@ -0,0 +1,375 @@
+#[derive(Clone, Eq)]
+pub struct Buffer<T: Bytes> {
+    boxed: Boxed<T>, // the boxed type of current GuardedVec
+}
```
```suggestion
/// [`Buffer`] is used when we want to store sensitive non encrypted data.
/// This shall always be short lived.
#[derive(Clone, Eq)]
pub struct Buffer<T: Bytes> {
    boxed: Boxed<T>, // the boxed type of current Buffer
}
```

elenaf9 replied:
Also, maybe re-add the old comment about the security features and guarantees?: 
```suggestion
/// [`Buffer`] is used when we want to store sensitive non encrypted data.
/// This shall always be short lived.
///
/// Provides the following features and guarantees:
/// * Causes segfault upon access without using a borrow.
/// * Protected using mprotect:
///    * `Prot::NoAccess` - when the box has no current borrows.
///    * `Prot::ReadOnly` - when the box has at least one current immutable borrow.
///    * `Prot::ReadWrite` - when the box has a current mutable borrow (can only have one at a time).
/// * The allocated memory uses guard pages both proceeding and following the memory. Overflows and large underflows
///   cause immediate termination of the program.
/// * A canary proceeds the memory location to detect smaller underflows.  The program will drop the underlying memory
///   and terminate if detected.
/// * The Memory is locked with `mlock`.
/// * When the memory is freed, `munlock` is called.
/// * The memory is zeroed when no longer in use via `sodium_free`.
/// * `Buffer` types can be compared in constant time.
/// * `Buffer` types can not be printed using `Debug`.
/// `Buffer` includes serialization which converts the data into a vector before its serialized by serde.  Upon
/// deserialization, the data is returned back to a new `Buffer`.
#[derive(Clone, Eq)]
pub struct Buffer<T: Bytes> {
    boxed: Boxed<T>, // the boxed type of current Buffer
}
```

elenaf9 commented on engine/new_runtime/src/memories/buffer.rs:
```diff
@@ -0,0 +1,375 @@
+        self.boxed.len()
+    }
+
+    #[allow(dead_code)]
```
Since `Buffer` is public I don't think we need  `#[allow(dead_code)]` for any of its methods.

elenaf9 commented on engine/new_runtime/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+[package]
```
A `.license_template` is missing for `new_runtime`.

elenaf9 commented on engine/new_runtime/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+features = [ "sha" ]
+
+[dev-dependencies]
+serde_json = "1.0"
```
```suggestion
serde_json = "1.0"

```
Nit: newline at end of file

elenaf9 commented on engine/new_runtime/src/memories/ram_memory.rs:
```diff
@@ -0,0 +1,159 @@
+/// is different, buffer type are meant for short lived usage while
+/// RamMemory can store data for longer period of time.
+/// Hence data in RamMemory has to be either encyrpted or protected
+/// behind a scheme
```
```suggestion
/// Protected ram memory that may be encrypted or not.
/// This is basically a wrapper for the [`Buffer`] type, but the usage
/// is different. The `Buffer` type is meant for short lived usage while
/// `RamMemory` can store data for longer period of time.
/// Hence data in `RamMemory` has to be either encrypted or protected
/// behind a scheme.
```

elenaf9 commented on engine/new_runtime/src/memories/ram_memory.rs:
```diff
@@ -0,0 +1,159 @@
+impl LockedMemory for RamMemory {
+    /// Locks the memory and possibly reallocates
+    // Currently we reallocate a new RamMemory at each lock
+    // This improves security but decreases performance
```
```suggestion
    /// Locks the memory and possibly reallocates.
    // Currently we reallocate a new [`RamMemory`] at each lock.
    // This improves security but decreases performance.
```

elenaf9 commented on engine/new_runtime/src/memories/ram_memory.rs:
```diff
@@ -0,0 +1,159 @@
+/// Hence data in RamMemory has to be either encyrpted or protected
+/// behind a scheme
+#[derive(Clone)]
+pub struct RamMemory {
```
```suggestion
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RamMemory {
```

No need to manually implement `Debug`, `Serialize` and `Deserialize` below. They are already implemented on `Buffer`, so we can just derive them here.

elenaf9 commented on engine/new_runtime/src/memories/ram_memory.rs:
```diff
@@ -0,0 +1,159 @@
+
+unsafe impl Send for RamMemory {}
+unsafe impl Sync for RamMemory {}
+
```
```suggestion
```

`Send` and `Sync` are automatically derived since `Buffer` already implements them. Quoting the Rustonomicon:

> Send and Sync are also automatically derived traits. This means that, unlike every other trait, if a type is composed entirely of Send or Sync types, then it is Send or Sync. Almost all primitives are Send and Sync, and as a consequence pretty much all types you'll ever interact with are Send and Sync.

elenaf9 commented on engine/new_runtime/src/memories/file_memory.rs:
```diff
@@ -0,0 +1,333 @@
+const FILENAME_SIZE: usize = 16;
+
+/// Data is stored into files in clear or encrypted.
+/// Basic security of this file includes files access control and
+pub struct FileMemory {
```
...and what? :D

tensor-programming replied:
haha

felsweg-iota replied:
you're so mean :rofl:

elenaf9 commented on engine/new_runtime/src/memories/noncontiguous_memory.rs:
```diff
@@ -0,0 +1,342 @@
+pub struct NonContiguousMemory {
+    shard1: MemoryShard,
+    shard2: MemoryShard,
+    config: NCConfig,
```
I don't think we need the `config` property here, it is only used on `alloc`. And if we need it again, we can just derive it by matching on the variants of `shard1` and `shard2`.

elenaf9 commented on engine/new_runtime/src/memories/noncontiguous_memory.rs:
```diff
@@ -0,0 +1,342 @@
+
+    // Refresh the shards to increase security, may be called every _n_ seconds or
+    // punctually
+    #[allow(dead_code)]
+    fn refresh(self) -> Result<Self, MemoryError> {
```
```suggestion
    pub fn refresh(self) -> Result<Self, MemoryError> {
```

Shouldn't this method be public to the user?

elenaf9 commented on engine/new_runtime/src/memories/noncontiguous_memory.rs:
```diff
@@ -0,0 +1,342 @@
+// Copyright 2020-2021 IOTA Stiftung
```
Amazing work with the non-contiguous memory!

elenaf9 commented on engine/src/vault/view.rs:
```diff
@@ -353,13 +353,20 @@ impl Record {
-            i.copy_from_slice(blob.as_ref());
-        });
+        let guarded = Buffer::alloc(&blob, tx.len.u64() as usize);
+
+        // let guarded =
+        //     GuardedVec::new(tx.len.u64() as usize, |i| {
+        //     let blob = SealedBlob::from(self.blob.as_ref())
+        //         .decrypt(key, tx.blob)
+        //         .expect("Unable to decrypt blob");
+
+        //     i.copy_from_slice(blob.as_ref());
+        // });
```
```suggestion
```

Remove old code.

elenaf9 commented on engine/src/vault/crypto_box.rs:
```diff
@@ -51,37 +57,32 @@ impl<T: BoxProvider> Key<T> {
             None
         }
     }
-
-    /// get the key's bytes from the [`GuardedVec`]
-    pub fn bytes(&self) -> Vec<u8> {
-        // hacks the guarded type.  Probably not the best solution.
-        (*self.key.borrow()).to_vec()
-    }
```
Why is this method removed? I think we need such a convenience method again, to avoid that the user has to the awkward call of:
```
let key = *(&key.key).borrow()
```
to access the actual key.

elenaf9 commented on engine/src/vault/crypto_box.rs:
```diff
@@ -36,11 +42,11 @@ pub trait BoxProvider: 'static + Sized + Ord + PartialOrd {
 pub struct Key<T: BoxProvider> {
     /// the guarded raw bytes that make up the key
-    pub key: GuardedVec<u8>,
+    pub key: Buffer<u8>,
```
I know in practice it does not make a difference, but should we use the `RamMemory` type here because the `Key` is stored for a longer period of time? I thought the `Buffer` type should only be used when accessing a value in a `LockedMemory`.

vuongDang replied:
Yes you're right, will change it

elenaf9 commented on engine/src/vault/crypto_box.rs:
```diff
@@ -149,3 +151,154 @@ pub trait Decrypt<T: TryFrom<Vec<u8>>>: AsRef<[u8]> {
+
+impl<T: BoxProvider> NCKey<T> {
+    /// generate a random key using secure random bytes
+    #[allow(dead_code)]
```
Shouldn't `engine/lib.rs` export this type? In which case we also don't need `#[allow(dead_code)]` on any of its public mehtods.

elenaf9 replied:
Resolved with #324.

elenaf9 commented on engine/src/vault/key_store.rs:
```diff
@@ -0,0 +1,91 @@
+};
+use std::collections::HashMap;
+
+/// The [`KeyStore`] keeps a map of [`VaultId`] -> [Vec<u8>] representing
+/// encrypted [`Key<P>`] using the `master_key`.
+/// `master_key` is stored in a non-contiguous data structure [`NCKey<P>`]
+/// for more security
+pub struct KeyStore<P: BoxProvider> {
+    store: HashMap<VaultId, Vec<u8>>,
+    master_key: NCKey<P>,
+}
```
Nice, this also solves issue #268!

elenaf9 commented on engine/src/vault/key_store.rs:
```diff
@@ -0,0 +1,91 @@
+}
+
+impl<P: BoxProvider> KeyStore<P> {
+    #![allow(dead_code)]
```
Again, I think `KeyStore` should also be public outside of the crate so we can use it in `iota_stronghold`. In that case we also don't need the `#![allow(dead_code)]`.

elenaf9 replied:
Resolved with #324.

elenaf9 reviewed (approved):
Really great work!

Sorry for my late review here, did not have time to look through #316 before it was merged.
Couple of comments. Almost all are nitpicks, the only blocking ones are that `engine::vault::{KeyStore, NCKey}` should be public if we want to use them in the client.

elenaf9 committed:
Feat/snapshot state in memory (#328)

* feat(client): secure snapshot state in memory

Implement `snapshot_state` RFC (Discussion#308).
This adds an additional layer of security for the snapshot state in
memory and allows to lazy load clients.

* test(client): test storing snapshot key

* tets(client): fix invalid snapshot file path

* feat(engine): make vault::{KeyStore, NCKey} public

* feat(client): use engine::KeyStore in snapshot

felsweg-iota committed:
Client Interface Update (#324)

* CI: add 'dev-refactor' branch for upcoming changes; add branch for CI to run; fix clippy errors

* fix: remove dev-refactor from covector version or publish

* CI: add 'dev-refactor' branch for upcoming changes; add branch for CI to run; fix clippy errors

* fix: remove dev-refactor from covector version or publish

* feat: primitives; blocker impl

* feat: first structure, but not functional

* feat: add benches; integration tests; integration of runtime / guarded mem as feature; removed old approach

* feat: remove read / write impl

* feat: blocking future without channels

* feat: remoe structural code

* feat: stm feature complete

* doc: add changes document

* feat: add composite key; extended doc

* fix: remove explicit typing of transaction

* feat: add stm crate to covector

* doc: fix doc

* regress: sync state, but non-functional. stm execution flaky in concurrent execution contexts.

* feat: read-log-update sync draft

* feat: rlu update draft #2

* feat: stricter atomic ops / tests still flaky

* feat: rlu skeleton impl

* feat: rlu reference impl

* feat: base api for rlu stm

* chore: remove old code

* feat: removed old code; add non-blocking queue / stack for write scaling

* feat: code cleanup / add tests

* feat: update tests, code cleanup

* merge: 2e62eaa2 from feat/locked-memory

* doc: fixes. minimal addition

* fix: clippy warnings; variant renaming; fix late initialization

* docfix: correct naming

* rename: stm -> rlu

* feat: factored out read/write traits for rlu to have functionality more configurable; nb tests still flaky.

* feat: cache proto impl for rlu showcasing

* fix: remove integration type proto

* feat: add breaker for rlu execution loop

* feat: add busy breaker for exponential retry handling

* fix: rluobject returns a reference to internal object

* fix: log clearing corrected index

* tests: add concurrent multiple writes / reads

* feat: replace write log with fixed sized array

* feat: restructure code.

* fix: code cleanup

* fix: factored out traits

* fix: factored out traits

* feat: factored out types

* feat: separate copy and var from enum

* feat: add spec dereference fn

* feat: replace raw pointer operations

* feat: remove raw pointers. working version.

* feat: sync state

* feat: add more tests

* fix: adapt date

* feat: add skeleton structure

* feat: add empty integration tests

* feat: add stub impl for new interface types

* feat: add cache with rlu

* feat: move cache skeleton to own module

* feat: add rlu cache proto impl

* feat: add tests for rlu cache

* feat: add concurrency tests

* fix: comment out current impl

* feat: rearranged interface types

* feat: add keyprovider stub

* feat: change store and client access

* feat: add provider; add keyprovider

* feat: Store type

* fix: add location types

* feat: migrated cryptographic procedures

* feat: migrated procedure runner implementation to client

* feat: move runner impl for client into extra module; impl procedure execution into Client

* feat: factored out thin client vault struct

* feat: remove outer mutability of procedures

* feat: remove unnecessary locks on types

* feat: make procedures immutable

* feat: examples

* fix: merge conflict; feat: add example README.md

* fix: date

* example: add BIP39

* feat: ported snapshot to new client api

* feat: add Stronghold root type

* fix: add cli tests for client_new

* fix: corrected keystore to get or insert key

elenaf9 committed:
feat(engine): low-level logic for merging vaults (#330)

Add fundamental low-level logic that can be used in client to sync
and merge vaults:
- list existing entries with id and blob-id
- changing encryption key and/ or id of a record
- exporting and importing records from / to a vault
- blob-id changes each time the data is updated

tensor-programming commented on client_new/src/types/stronghold/base.rs:
```diff
@@ -0,0 +1,192 @@
+use engine::vault::ClientId;
+use std::ops::Deref;
+
+impl Stronghold {
```
I know these aren't the complete interfaces for the user to interact with, but maybe it would be a little bit more clear how a user should interact with this library if the stronghold type and interfaces weren't nested so far in a folder like this.  The same for the other front facing types, client, keystore etc.  Or at least some docs that highlight that they are the main types that should be used.

felsweg-iota replied:
yeah, docs coming up. i mean for the `Stronghold` type the intention was to separate `base` impl. for all local  and `p2p` for all (upcoming / ported) networking functionality.

tensor-programming commented on client_new/src/tests/interface_tests.rs:
```diff
@@ -0,0 +1,118 @@
+// Copyright 2020-2021 IOTA Stiftung
```
I wonder, do these tests even need to be a part of the core program anymore or can they be put into a test file?  The reason before was because there were some items things that weren't public which needed to be tested through the interface.

felsweg-iota replied:
could be moved into an integration tests, yes

felsweg-iota committed:
doc: update on front facing README / CONTRIBUTION guide.

felsweg-iota committed:
Update README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

felsweg-iota committed:
Update README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

felsweg-iota committed:
Update README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

felsweg-iota committed:
Update README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

felsweg-iota committed:
Update README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

felsweg-iota committed:
Update README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

felsweg-iota committed:
Update README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

felsweg-iota committed:
Update README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

felsweg-iota committed:
Update README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

felsweg-iota committed:
Update CONTRIBUTING.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

elenaf9 commented on client_new/src/types/client.rs:
````diff
@@ -0,0 +1,207 @@
+        self.store.atomic_ref()
+    }
+
+    /// Returns a [`Vault`] according to path
+    ///
+    /// # Example
+    /// ```
+    /// ```
+    pub fn vault(&self, vault_path: Location) -> ClientVault {
+        let (vault_id, _) = vault_path.resolve();
+
+        ClientVault {
+            client: self.atomic_ref(),
+            id: vault_id,
+        }
+    }
````
With #319 all operations on the vault can be done through procedures. We could consider removing `ClientVault` so that we have a single way of doing things through procedures only.

felsweg-iota replied:
`ClientVault` is a thin wrapper around `Client` and calls all actions directly on `Client`s database.

elenaf9 replied:
Yes but said actions (`write_secret`, `delete_secret`, `revoke_secret` and `cleanup`) can all be done through procedures (`StrongholdProcedure::WriteVault`, `StrongholdProcedure::RevokeData`, etc). 
Thus the methods are kinda redundant and if we would remove them, we also would not need this `ClientVault` wrapper anymore.

felsweg-iota replied:
I disagree. The `ClientVault` is meant to have as a distinct type to differentiate the actions taken on them. The whole idea of the new client interface is around types.

felsweg-iota replied:
The only point to argue would be to expose this `ClientVault` as an interface on `Client`.

elenaf9 replied:
My argument was based on what is described in the _Primitive Based Interface_ RFC in #313 in **Reference-level explanation - Vault**: 

> While the vault is a stronghold primitive, it does not have its own type (being the only exception). Most interesting operations on a vault are procedures. Since a client can have multiple vaults, a procedure generally works on the client-level rather than the vault-level (...)

I think it's cleaner to have a unified way of interacting with vaults by always using procedures. It's also easier for the user to understand, rather than having multiple ways of how something can be written to a vault.

But I also see your point. Having a `ClientVault` type is more consistent with how the interface is generally designed.

Anyway, your call 😄.

felsweg-iota replied:
The execution of procedures are on `Client` level, while `ClientVault` offers a more distinct type :D Sure, every action can be done by procedures, but having access to "simpler" function calls might be beneficial, imho.

elenaf9 commented on client_new/src/types/client.rs:
````diff
@@ -0,0 +1,207 @@
+    /// # Example
+    /// ```no_run
+    /// ```
+    pub async fn execure_procedure_chained(
````
Why are some of the methods on the `Client` async? Unless I am missing something, there is no need for it.

felsweg-iota replied:
true, it's a reminiscent of the old interface. Most of the operations are sync, since we do not enforce a hard dependency on some async runtime. I/O operations on the `Snapshot` should be async, imho. I wonder, if we should enforce a dependency on tokio (or any other async runtime)

elenaf9 committed:
Feat(client_new): State synchronization of Clients and Snapshots (#334)

* feat(client): implement sync for clients

* test(client): test client-sync

* feat(client): implements sync for snapshots

* fix(client): remove unnecessary mut access

* refactor(client): improve sync, enable p2p-sync

* feat(client_new): handle errors

* feat(client_new): use path instead of id in interface

* fix(client_new): AsRef<[u8]> instead of <Vec<u8>>

* feat(client_new): use generics for paths in sync

felsweg-iota committed:
examples/client interface (#333)

* examples: extended example

* feat: completed docs

* Update client_new/examples/README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

* Update client_new/examples/README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

* Update client_new/examples/README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

* Update client_new/examples/README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

* Update client_new/examples/README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

* Update client_new/examples/README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

* Update client_new/examples/README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

* Update client_new/examples/README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

* Update client_new/examples/README.md

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

* fix: add client_new to ci (for now)

* docfix: put example sections as sublists

Co-authored-by: Lucas Tortora <85233773+lucas-tortora@users.noreply.github.com>

felsweg-iota committed:
feat: ported network skeleton

felsweg-iota committed:
merge dev-refactor

felsweg-iota committed:
feat: minimal network client port

felsweg-iota committed:
feat: add Peer type as thin abstraction over remote clients

felsweg-iota committed:
feat: extended tests; store operations

Dr-Electron committed:
Prepare for new wiki build system

felsweg-iota committed:
feat: write / read to remote store working version

felsweg-iota committed:
feat: delete from store

felsweg-iota committed:
fix: remove unused tests

felsweg-iota committed:
feat: feature gate network tests

felsweg-iota committed:
fix: ignore doc tests

felsweg-iota committed:
feat: add generate key / remote procedure call

felsweg-iota committed:
feat: take location as ref

felsweg-iota committed:
feat: remove 'RecordHint' from active usage

felsweg-iota committed:
feat: derive Display for SnapshotPath; GuardDebug for KeyProvider and Stronghold

Dr-Electron committed:
Update to yarn 3.2.0 and add wiki cli

felsweg-iota committed:
feat: ported remote synchronization

JSto91 committed:
[Wiki] Add Banner

felsweg-iota committed:
feat: removed unused code

felsweg-iota committed:
fix: code cleanup

felsweg-iota committed:
Merge pull request #337 from iotaledger/feat/network-client-port

Network Client Port (Sync current impl)

felsweg-iota committed:
doc: fix docs

felsweg-iota committed:
chore: remove old crates

felsweg-iota committed:
fix: workflows

felsweg-iota committed:
chore: remote commandline product. see client/examples for related product

felsweg-iota committed:
fix: tests import renamed to iota_stronghold

felsweg-iota committed:
fix: set file to write enabled

felsweg-iota committed:
Merge pull request #346 from iotaledger/chore/move-updated-implementation

fix: remove old client interface, old runtime

PhilippGackstatter committed:
Implement concat kdf procedure

felsweg-iota committed:
fix: remove naive kdf; replace sha256 with blake2b256

PhilippGackstatter committed:
Merge branch 'dev-refactor' into feat/concat-kdf

PhilippGackstatter committed:
Apply fmt

felsweg-iota committed:
doc: fix intradoc links

felsweg-iota committed:
doc: fix intradoc links

felsweg-iota committed:
Merge pull request #349 from iotaledger/chore/fix-documentation-links

Fix documentation links

JSto91 committed:
Update README.md

PhilippGackstatter committed:
Add test vector for concat kdf

PhilippGackstatter committed:
Document algorithm; add pub/prv info params

PhilippGackstatter committed:
Simplify digest generics

PhilippGackstatter committed:
Set `ClientId` when loading a client

PhilippGackstatter committed:
Remove unnecessary imports and variables

PhilippGackstatter committed:
Merge pull request #347 from iotaledger/feat/concat-kdf

Add the `ConcatKdf` procedure

PhilippGackstatter committed:
Merge pull request #351 from iotaledger/fix/set-id-load-client

Fix `ClientId` not being set when loading clients

PhilippGackstatter committed:
Allow relative paths in `SnapshotPath`

PhilippGackstatter committed:
Merge pull request #352 from iotaledger/fix/allow-relative-paths

Allow relative paths for snapshots

lmoe committed:
Initial bindings implementation

lmoe committed:
Add signing

lmoe committed:
remove db

lmoe committed:
Cleanup

lmoe committed:
Remove archive

lmoe committed:
Add module wrapper, tests

lmoe committed:
Add cbindgen, update go lib/tests

lmoe committed:
Add seed generation/derive, get address

felsweg-iota committed:
Merge branch 'dev-refactor' into doc/README-update

felsweg-iota committed:
Merge remote-tracking branch 'refs/remotes/origin/doc/README-update' into doc/README-update

felsweg-iota committed:
doc: add security audit statement of work performed

PhilippGackstatter committed:
Allow procedures to use multiple secrets

lmoe committed:
Refactor library

lmoe committed:
Refactoring

PhilippGackstatter committed:
Allow exec_proc to use multiple source locations

PhilippGackstatter committed:
Fix procedures macros; introduce type alias

PhilippGackstatter committed:
Document methods

PhilippGackstatter committed:
Use explicit `try_from` as opposed to `try_into`

PhilippGackstatter committed:
Fix incorrect variable name in expect

felsweg-iota committed:
Update CONTRIBUTING.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update CONTRIBUTING.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update CONTRIBUTING.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update CONTRIBUTING.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update CONTRIBUTING.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update CONTRIBUTING.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update CONTRIBUTING.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update README.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update README.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update CONTRIBUTING.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
fix: ported procedures tests; fixes

felsweg-iota committed:
fix: add hashed_secret test

felsweg-iota committed:
Merge pull request #348 from iotaledger/security/replace-hashing-function-for-nc-types

Remove `naiveKDF` function; Fix digest for NCTypes

PhilippGackstatter committed:
Merge pull request #354 from iotaledger/feat/multiple-secrets-procedures

 Allow procedures to use multiple secrets

Dr-Electron committed:
Cleanup PR and update cli

Dr-Electron committed:
Add test action

Dr-Electron committed:
Update docusaurus.config.js

lmoe committed:
Refactoring, completion of first stronghold native alpha

lmoe committed:
Add license remarks

lmoe committed:
Refactoring

lmoe committed:
Update code according to PR comments

lmoe committed:
Update derived chain code, fix typo

lmoe committed:
Merge pull request #356 from lmoe/add_native_bindings

Add native bindings

Dr-Electron committed:
Fix wording

felsweg-iota committed:
Merge pull request #332 from Dr-Electron/docs/new-build-system

Prepare for new wiki build system

felsweg-iota committed:
doc: add Stronghold Enhancement Proposal Process

felsweg-iota committed:
Merge pull request #345 from JSto91/dev

[Wiki] Add Banner

felsweg-iota committed:
doc: merge wiki banner changes into README

felsweg-iota committed:
doc: merge wiki banner changes into README

felsweg-iota committed:
warn: allow non-async mutex across wait points.

felsweg-iota committed:
Merge branch 'dev-refactor' into doc/README-update

felsweg-iota committed:
Update README.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update SEP.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update SEP.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update SEP.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update SEP.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update SEP.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update SEP.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update SEP.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update SEP.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

felsweg-iota committed:
Update SEP.md

Co-authored-by: charlesthompson3 <74603461+charlesthompson3@users.noreply.github.com>

Dr-Electron committed:
Trigger broken build

Dr-Electron committed:
Add static img path

Dr-Electron committed:
Fix branch name

Dr-Electron committed:
Update wiki cli

Dr-Electron committed:
Fix static img path (#357)

PhilippGackstatter committed:
Expose procedures for AES Key Wrap (#355)

* Implement AES256KW encryption procedure

* Use generic_procedures macro

* Document ciphers; fix ctx output length

* Implement the `AesKeyWrapDecrypt` procedure

* Return error for ciphertext of len < BLOCK

felsweg-iota committed:
feat: Enable Clone on keys for Cache; list all keys inside Store

felsweg-iota committed:
fix: fix doctest

felsweg-iota committed:
fix: fix doctest

felsweg-iota committed:
fix: doctests, also renamed to keys() for Cache

felsweg-iota committed:
Merge pull request #359 from iotaledger/feat/list-store-keys

Function to list keys for Store

felsweg-iota committed:
chore(release-doc): clean up outdated release notes, merge existing into one

felsweg-iota committed:
Merge pull request #331 from iotaledger/doc/README-update

doc: update on front-facing README / CONTRIBUTION guide.

felsweg-iota committed:
doc: remove duplicate jump

felsweg-iota committed:
fix: ignore network synchronization

felsweg-iota committed:
fix: add more tests

felsweg-iota committed:
Merge pull request #361 from iotaledger/fixes/provide-additional-tests

Additional interface tests

felsweg-iota committed:
Merge pull request #360 from iotaledger/chore/release-docs

chore(release-doc): clean up outdated release docs / merge existing
